// @mui/joy/modern/Select/Select.js
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";
import _extends from "@babel/runtime/helpers/esm/extends";

var _Unfold;

const _excluded = ["action", "autoFocus", "children", "componentsProps", "defaultValue", "defaultListboxOpen", "disabled", "getSerializedValue", "placeholder", "listboxId", "listboxOpen", "onChange", "onListboxOpenChange", "onClose", "renderValue", "value", "size", "variant", "color", "startDecorator", "endDecorator", "indicator", "aria-describedby", "aria-label", "aria-labelledby", "id", "name"],
      _excluded2 = ["component"];
import * as React from 'react';
import PropTypes from 'prop-types';
import { unstable_capitalize as capitalize, unstable_useForkRef as useForkRef, unstable_useControlled as useControlled } from '@mui/utils';
import PopperUnstyled from '@mui/base/PopperUnstyled';
import { useSelect, SelectUnstyledContext, flattenOptionGroups, getOptionsFromChildren } from '@mui/base/SelectUnstyled';
import { useSlotProps } from '@mui/base/utils';
import composeClasses from '@mui/base/composeClasses';
import { ListRoot } from '../List/List';
import ListProvider, { scopedVariables } from '../List/ListProvider';
import Unfold from '../internal/svg-icons/Unfold';
import { styled, useThemeProps } from '../styles';
import selectClasses, { getSelectUtilityClass } from './selectClasses';
import FormControlContext from '../FormControl/FormControlContext';
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";

function defaultRenderSingleValue(selectedOption) {
  return selectedOption?.label ?? '';
}

function defaultFormValueProvider(selectedOption) {
  if (selectedOption?.value == null) {
    return '';
  }

  if (typeof selectedOption.value === 'string' || typeof selectedOption.value === 'number') {
    return selectedOption.value;
  }

  return JSON.stringify(selectedOption.value);
}

const useUtilityClasses = ownerState => {
  const {
    color,
    disabled,
    focusVisible,
    size,
    variant,
    open
  } = ownerState;
  const slots = {
    root: ['root', disabled && 'disabled', focusVisible && 'focusVisible', open && 'expanded', variant && `variant${capitalize(variant)}`, color && `color${capitalize(color)}`, size && `size${capitalize(size)}`],
    button: ['button'],
    startDecorator: ['startDecorator'],
    endDecorator: ['endDecorator'],
    indicator: ['indicator', open && 'expanded'],
    listbox: ['listbox', open && 'expanded', disabled && 'disabled', variant && `variant${capitalize(variant)}`, color && `color${capitalize(color)}`, size && `size${capitalize(size)}`]
  };
  return composeClasses(slots, getSelectUtilityClass, {});
};

const SelectRoot = styled('div', {
  name: 'JoySelect',
  slot: 'Root',
  overridesResolver: (props, styles) => styles.root
})(({
  theme,
  ownerState
}) => {
  const variantStyle = theme.variants[`${ownerState.variant}`]?.[ownerState.color];
  return [_extends({
    '--focus-outline-offset': `calc(${theme.vars.focus.thickness} * -1)`,
    // to prevent the focus outline from being cut by overflow
    '--Select-radius': theme.vars.radius.sm,
    '--Select-gap': '0.5rem',
    '--Select-placeholderOpacity': 0.5,
    '--Select-focusedThickness': theme.vars.focus.thickness,
    '--Select-focusedHighlight': theme.vars.palette[ownerState.color === 'neutral' ? 'primary' : ownerState.color]?.[500],
    '--Select-indicator-color': theme.vars.palette.text.tertiary
  }, ownerState.size === 'sm' && {
    '--Select-minHeight': '2rem',
    '--Select-paddingInline': '0.5rem',
    '--Select-decorator-childHeight': 'min(1.5rem, var(--Select-minHeight))',
    '--Icon-fontSize': '1.25rem'
  }, ownerState.size === 'md' && {
    '--Select-minHeight': '2.5rem',
    '--Select-paddingInline': '0.75rem',
    '--Select-decorator-childHeight': 'min(2rem, var(--Select-minHeight))',
    '--Icon-fontSize': '1.5rem'
  }, ownerState.size === 'lg' && {
    '--Select-minHeight': '3rem',
    '--Select-paddingInline': '1rem',
    '--Select-decorator-childHeight': 'min(2.375rem, var(--Select-minHeight))',
    '--Icon-fontSize': '1.75rem'
  }, {
    // variables for controlling child components
    '--Select-decorator-childOffset': 'min(calc(var(--Select-paddingInline) - (var(--Select-minHeight) - 2 * var(--variant-borderWidth) - var(--Select-decorator-childHeight)) / 2), var(--Select-paddingInline))',
    '--internal-paddingBlock': 'max((var(--Select-minHeight) - 2 * var(--variant-borderWidth) - var(--Select-decorator-childHeight)) / 2, 0px)',
    '--Select-decorator-childRadius': 'max(var(--Select-radius) - var(--internal-paddingBlock), min(var(--internal-paddingBlock) / 2, var(--Select-radius) / 2))',
    '--Button-minHeight': 'var(--Select-decorator-childHeight)',
    '--IconButton-size': 'var(--Select-decorator-childHeight)',
    '--Button-radius': 'var(--Select-decorator-childRadius)',
    '--IconButton-radius': 'var(--Select-decorator-childRadius)',
    boxSizing: 'border-box',
    minWidth: 0,
    minHeight: 'var(--Select-minHeight)',
    position: 'relative',
    display: 'flex',
    alignItems: 'center',
    borderRadius: 'var(--Select-radius)'
  }, !variantStyle.backgroundColor && {
    backgroundColor: theme.vars.palette.background.surface
  }, {
    paddingInline: `var(--Select-paddingInline)`,
    fontFamily: theme.vars.fontFamily.body,
    fontSize: theme.vars.fontSize.md
  }, ownerState.size === 'sm' && {
    fontSize: theme.vars.fontSize.sm
  }, {
    // TODO: discuss the transition approach in a separate PR.
    transition: 'background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    '&:before': {
      boxSizing: 'border-box',
      content: '""',
      display: 'block',
      position: 'absolute',
      pointerEvents: 'none',
      top: 0,
      left: 0,
      right: 0,
      bottom: 0,
      zIndex: 1,
      borderRadius: 'inherit',
      margin: 'calc(var(--variant-borderWidth) * -1)' // for outlined variant

    },
    [`&.${selectClasses.focusVisible}`]: {
      '--Select-indicator-color': 'var(--Select-focusedHighlight)'
    },
    [`&.${selectClasses.disabled}`]: {
      '--Select-indicator-color': 'inherit'
    }
  }, ownerState.variant !== 'solid' && {
    [`&.${selectClasses.focusVisible}`]: {
      '&:before': {
        boxShadow: `inset 0 0 0 var(--Select-focusedThickness) var(--Select-focusedHighlight)`
      }
    }
  }), _extends({}, variantStyle, {
    '&:hover': theme.variants[`${ownerState.variant}Hover`]?.[ownerState.color],
    [`&.${selectClasses.disabled}`]: theme.variants[`${ownerState.variant}Disabled`]?.[ownerState.color]
  })];
});
const SelectButton = styled('button', {
  name: 'JoySelect',
  slot: 'Button',
  overridesResolver: (props, styles) => styles.button
})(({
  ownerState
}) => _extends({
  // reset user-agent button style
  border: 0,
  outline: 'none',
  background: 'none',
  padding: 0,
  fontSize: 'inherit',
  color: 'inherit',
  alignSelf: 'stretch',
  // make children horizontally aligned
  display: 'flex',
  alignItems: 'center',
  flex: 1,
  cursor: 'pointer'
}, (ownerState.value === null || ownerState.value === undefined) && {
  opacity: 'var(--Select-placeholderOpacity)'
}));
const SelectListbox = styled(ListRoot, {
  name: 'JoySelect',
  slot: 'Listbox',
  overridesResolver: (props, styles) => styles.listbox
})(({
  theme,
  ownerState
}) => {
  const variantStyle = theme.variants[ownerState.variant]?.[ownerState.color];
  return _extends({
    '--List-radius': theme.vars.radius.sm,
    '--List-item-stickyBackground': variantStyle?.backgroundColor || variantStyle?.background || theme.vars.palette.background.surface,
    // for sticky List
    '--List-item-stickyTop': 'calc(var(--List-padding, var(--List-divider-gap)) * -1)'
  }, scopedVariables, {
    outline: 'none',
    boxShadow: theme.vars.shadow.md,
    zIndex: 1000
  }, !variantStyle.backgroundColor && {
    backgroundColor: theme.vars.palette.background.surface
  });
});
const SelectStartDecorator = styled('span', {
  name: 'JoySelect',
  slot: 'StartDecorator',
  overridesResolver: (props, styles) => styles.startDecorator
})(({
  theme,
  ownerState
}) => _extends({
  '--Button-margin': '0 0 0 calc(var(--Select-decorator-childOffset) * -1)',
  '--IconButton-margin': '0 0 0 calc(var(--Select-decorator-childOffset) * -1)',
  '--Icon-margin': '0 0 0 calc(var(--Select-paddingInline) / -4)',
  pointerEvents: 'none',
  // to make the input focused when click on the element because start element usually is an icon
  display: 'inherit',
  alignItems: 'center',
  marginInlineEnd: 'var(--Select-gap)',
  color: theme.vars.palette.text.tertiary
}, ownerState.focusVisible && {
  color: 'var(--Select-focusedHighlight)'
}));
const SelectEndDecorator = styled('span', {
  name: 'JoySelect',
  slot: 'EndDecorator',
  overridesResolver: (props, styles) => styles.endDecorator
})(({
  theme,
  ownerState
}) => ({
  '--Button-margin': '0 calc(var(--Select-decorator-childOffset) * -1) 0 0',
  '--IconButton-margin': '0 calc(var(--Select-decorator-childOffset) * -1) 0 0',
  '--Icon-margin': '0 calc(var(--Select-paddingInline) / -4) 0 0',
  display: 'inherit',
  alignItems: 'center',
  marginInlineStart: 'var(--Select-gap)',
  color: theme.vars.palette[ownerState.color]?.[`${ownerState.variant}Color`]
}));
const SelectIndicator = styled('span', {
  name: 'JoySelect',
  slot: 'Indicator'
})(({
  ownerState
}) => _extends({}, ownerState.size === 'sm' && {
  '--Icon-fontSize': '1.125rem'
}, ownerState.size === 'md' && {
  '--Icon-fontSize': '1.25rem'
}, ownerState.size === 'lg' && {
  '--Icon-fontSize': '1.5rem'
}, {
  color: 'var(--Select-indicator-color)',
  display: 'inherit',
  alignItems: 'center',
  marginInlineStart: 'var(--Select-gap)',
  marginInlineEnd: 'calc(var(--Select-paddingInline) / -4)',
  [`.${selectClasses.endDecorator} + &`]: {
    marginInlineStart: 'calc(var(--Select-gap) / 2)'
  }
}));
const Select = /*#__PURE__*/React.forwardRef(function Select(inProps, ref) {
  const props = useThemeProps({
    props: inProps,
    name: 'JoySelect'
  });

  const _ref = props,
        {
    action,
    autoFocus,
    children,
    componentsProps = {},
    defaultValue,
    defaultListboxOpen = false,
    disabled: disabledExternalProp,
    getSerializedValue = defaultFormValueProvider,
    placeholder,
    listboxId,
    listboxOpen: listboxOpenProp,
    onChange,
    onListboxOpenChange,
    onClose,
    renderValue: renderValueProp,
    value: valueProp,
    size: sizeProp = 'md',
    variant = 'outlined',
    color: colorProp = 'neutral',
    startDecorator,
    endDecorator,
    indicator = _Unfold || (_Unfold = /*#__PURE__*/_jsx(Unfold, {})),
    // props to forward to the button (all handlers should go through componentsProps.button)
    'aria-describedby': ariaDescribedby,
    'aria-label': ariaLabel,
    'aria-labelledby': ariaLabelledby,
    id,
    name
  } = _ref,
        other = _objectWithoutPropertiesLoose(_ref, _excluded);

  const formControl = React.useContext(FormControlContext);

  if (process.env.NODE_ENV !== 'production') {
    const registerEffect = formControl?.registerEffect; // eslint-disable-next-line react-hooks/rules-of-hooks

    React.useEffect(() => {
      if (registerEffect) {
        return registerEffect();
      }

      return undefined;
    }, [registerEffect]);
  }

  const disabledProp = inProps.disabled ?? formControl?.disabled ?? disabledExternalProp;
  const size = inProps.size ?? formControl?.size ?? sizeProp;
  const color = formControl?.error ? 'danger' : inProps.color ?? formControl?.color ?? colorProp;
  const renderValue = renderValueProp ?? defaultRenderSingleValue;
  const [anchorEl, setAnchorEl] = React.useState(null);
  const [groupedOptions, setGroupedOptions] = React.useState([]);
  const options = React.useMemo(() => flattenOptionGroups(groupedOptions), [groupedOptions]);
  const [listboxOpen, setListboxOpen] = useControlled({
    controlled: listboxOpenProp,
    default: defaultListboxOpen,
    name: 'SelectUnstyled',
    state: 'listboxOpen'
  });
  const rootRef = React.useRef(null);
  const buttonRef = React.useRef(null);
  const listboxRef = React.useRef(null);
  const handleRef = useForkRef(ref, rootRef);
  React.useImperativeHandle(action, () => ({
    focusVisible: () => {
      buttonRef.current?.focus();
    }
  }), []);
  React.useEffect(() => {
    setGroupedOptions(getOptionsFromChildren(children));
  }, [children]);
  React.useEffect(() => {
    setAnchorEl(rootRef.current);
  }, []);
  React.useEffect(() => {
    if (autoFocus) {
      buttonRef.current.focus();
    }
  }, [autoFocus]);

  const handleOpenChange = isOpen => {
    setListboxOpen(isOpen);
    onListboxOpenChange?.(isOpen);

    if (!isOpen) {
      onClose?.();
    }
  };

  const {
    buttonActive,
    buttonFocusVisible,
    disabled,
    getButtonProps,
    getListboxProps,
    getOptionProps,
    getOptionState,
    value
  } = useSelect({
    buttonRef,
    defaultValue,
    disabled: disabledProp,
    listboxId,
    multiple: false,
    onChange,
    onOpenChange: handleOpenChange,
    open: listboxOpen,
    options,
    value: valueProp
  });

  const ownerState = _extends({}, props, {
    active: buttonActive,
    defaultListboxOpen,
    disabled,
    focusVisible: buttonFocusVisible,
    open: listboxOpen,
    renderValue,
    value,
    size,
    variant,
    color
  });

  const classes = useUtilityClasses(ownerState);
  const selectedOption = React.useMemo(() => {
    return options.find(o => value === o.value) ?? null;
  }, [options, value]);
  const rootProps = useSlotProps({
    elementType: SelectRoot,
    getSlotProps: handlers => ({
      onMouseDown: event => {
        if (!listboxOpen && event.target !== buttonRef.current && !event.isPropagationStopped()) {
          // show the popup if user click outside of the button element.
          // the close action is already handled by blur event.
          handleOpenChange(true);
        }

        handlers.onClick?.(event);
      }
    }),
    externalSlotProps: componentsProps.root,
    externalForwardedProps: other,
    additionalProps: {
      ref: handleRef
    },
    ownerState,
    className: classes.root
  });
  const buttonProps = useSlotProps({
    elementType: SelectButton,
    getSlotProps: getButtonProps,
    externalSlotProps: componentsProps.button,
    additionalProps: {
      'aria-describedby': ariaDescribedby ?? formControl?.['aria-describedby'],
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledby ?? formControl?.labelId,
      id: id ?? formControl?.htmlFor,
      name
    },
    ownerState,
    className: classes.button
  });
  const resolveListboxProps = typeof componentsProps.listbox === 'function' ? componentsProps.listbox(ownerState) : componentsProps.listbox; // cache the modifiers to prevent Popper from being recreated when React rerenders menu.

  const cachedModifiers = React.useMemo(() => [{
    name: 'offset',
    options: {
      offset: [0, 4]
    }
  }, {
    // popper will have the same width as root element when open
    name: 'equalWidth',
    enabled: true,
    phase: 'beforeWrite',
    requires: ['computeStyles'],
    fn: ({
      state
    }) => {
      state.styles.popper.width = `${state.rects.reference.width}px`;
    }
  }, ...(resolveListboxProps?.modifiers || [])], [resolveListboxProps?.modifiers]);

  const _useSlotProps = useSlotProps({
    elementType: SelectListbox,
    getSlotProps: getListboxProps,
    externalSlotProps: componentsProps.listbox,
    additionalProps: {
      ref: listboxRef,
      anchorEl,
      disablePortal: true,
      open: listboxOpen,
      placement: 'bottom',
      modifiers: cachedModifiers
    },
    ownerState: _extends({}, ownerState, {
      nesting: false,
      row: false,
      wrap: false
    }),
    className: classes.listbox
  }),
        {
    component: listboxComponent
  } = _useSlotProps,
        listboxProps = _objectWithoutPropertiesLoose(_useSlotProps, _excluded2);

  const startDecoratorProps = useSlotProps({
    elementType: SelectStartDecorator,
    externalSlotProps: componentsProps.startDecorator,
    ownerState,
    className: classes.startDecorator
  });
  const endDecoratorProps = useSlotProps({
    elementType: SelectEndDecorator,
    externalSlotProps: componentsProps.endDecorator,
    ownerState,
    className: classes.endDecorator
  });
  const indicatorProps = useSlotProps({
    elementType: SelectIndicator,
    externalSlotProps: componentsProps.indicator,
    ownerState,
    className: classes.indicator
  });
  const context = React.useMemo(() => ({
    getOptionProps,
    getOptionState,
    listboxRef,
    color
  }), [color, getOptionProps, getOptionState]);
  return /*#__PURE__*/_jsxs(React.Fragment, {
    children: [/*#__PURE__*/_jsxs(SelectRoot, _extends({}, rootProps, {
      children: [startDecorator && /*#__PURE__*/_jsx(SelectStartDecorator, _extends({}, startDecoratorProps, {
        children: startDecorator
      })), /*#__PURE__*/_jsx(SelectButton, _extends({}, buttonProps, {
        children: selectedOption ? renderValue(selectedOption) : placeholder
      })), endDecorator && /*#__PURE__*/_jsx(SelectEndDecorator, _extends({}, endDecoratorProps, {
        children: endDecorator
      })), indicator && /*#__PURE__*/_jsx(SelectIndicator, _extends({}, indicatorProps, {
        children: indicator
      }))]
    })), anchorEl &&
    /*#__PURE__*/
    // @ts-ignore internal logic: `listboxComponent` should not replace `SelectListbox`.
    _jsx(PopperUnstyled, _extends({}, listboxProps, {
      as: listboxComponent,
      component: SelectListbox,
      children: /*#__PURE__*/_jsx(SelectUnstyledContext.Provider, {
        value: context,
        children: /*#__PURE__*/_jsx(ListProvider, {
          nested: true,
          children: children
        })
      })
    })), name && /*#__PURE__*/_jsx("input", {
      type: "hidden",
      name: name,
      value: getSerializedValue(selectedOption)
    })]
  });
});
process.env.NODE_ENV !== "production" ? Select.propTypes
/* remove-proptypes */
= {
  // ----------------------------- Warning --------------------------------
  // | These PropTypes are generated from the TypeScript type definitions |
  // |     To update them edit TypeScript types and run "yarn proptypes"  |
  // ----------------------------------------------------------------------

  /**
   * A ref for imperative actions. It currently only supports `focusVisible()` action.
   */
  action: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({
    current: PropTypes.shape({
      focusVisible: PropTypes.func.isRequired
    })
  })]),

  /**
   * @ignore
   */
  children: PropTypes.node,

  /**
   * The color of the component. It supports those theme colors that make sense for this component.
   * @default 'primary'
   */
  color: PropTypes
  /* @typescript-to-proptypes-ignore */
  .oneOfType([PropTypes.oneOf(['danger', 'info', 'neutral', 'primary', 'success', 'warning']), PropTypes.string]),

  /**
   * The component used for the root node.
   * Either a string to use a HTML element or a component.
   */
  component: PropTypes.elementType,

  /**
   * The props used for each slot inside the component.
   * @default {}
   */
  componentsProps: PropTypes.shape({
    button: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    endDecorator: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    indicator: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    listbox: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    root: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
    startDecorator: PropTypes.oneOfType([PropTypes.func, PropTypes.object])
  }),

  /**
   * The default selected value. Use when the component is not controlled.
   */
  defaultValue: PropTypes.any,

  /**
   * If `true`, the component is disabled.
   * @default false
   */
  disabled: PropTypes.bool,

  /**
   * Trailing adornment for the select.
   */
  endDecorator: PropTypes.node,

  /**
   * A function to convert the currently selected value to a string.
   * Used to set a value of a hidden input associated with the select,
   * so that the selected value can be posted with a form.
   */
  getSerializedValue: PropTypes.func,

  /**
   * The indicator(*) for the select.
   *    ________________
   *   [ value        * ]
   *    ‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾‾
   */
  indicator: PropTypes.node,

  /**
   * Callback fired when an option is selected.
   */
  onChange: PropTypes.func,

  /**
   * Triggered when focus leaves the menu and the menu should close.
   */
  onClose: PropTypes.func,

  /**
   * Text to show when there is no selected value.
   */
  placeholder: PropTypes.node,

  /**
   * Function that customizes the rendering of the selected value.
   */
  renderValue: PropTypes.func,

  /**
   * The size of the component.
   */
  size: PropTypes
  /* @typescript-to-proptypes-ignore */
  .oneOfType([PropTypes.oneOf(['sm', 'md', 'lg']), PropTypes.string]),

  /**
   * Leading adornment for the select.
   */
  startDecorator: PropTypes.node,

  /**
   * The system prop that allows defining system overrides as well as additional CSS styles.
   */
  sx: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.func, PropTypes.object, PropTypes.bool])), PropTypes.func, PropTypes.object]),

  /**
   * The selected value.
   * Set to `null` to deselect all options.
   */
  value: PropTypes.any,

  /**
   * The variant to use.
   * @default 'solid'
   */
  variant: PropTypes
  /* @typescript-to-proptypes-ignore */
  .oneOfType([PropTypes.oneOf(['outlined', 'plain', 'soft', 'solid']), PropTypes.string])
} : void 0;
export default Select;
